subprojects/platform-native/src/main/groovy/org/gradle/nativeplatform/toolchain/internal/msvcpp/DefaultUcrtLocator.java

```diff
@@ -0,0 +1,245 @@
+                } else {
+                    LOGGER.debug("Ignoring candidate ucrt directory {} as it does not look like a ucrt installation.", ucrtDir);
+                }
+            } catch (MissingRegistryEntryException e) {
+                // Ignore the version if the string cannot be read
+            }
+        }
+    }
+
+    private SearchResult locateUserSpecifiedUcrt(File candidate) {
+        File ucrtDir = GFileUtils.canonicalise(candidate);
+        String[] versionDirs = getUcrtVersionDirs(ucrtDir);
+        if (versionDirs.length > 0) {
+            for (String versionDir : versionDirs) {
+                LOGGER.debug("Found ucrt {} ({}) at {}", VERSION_USER, versionDir, ucrtDir);
+                if (!foundUcrts.containsKey(ucrtDir)) {
+                    putUcrt(new Ucrt(ucrtDir, VersionNumber.parse(VERSION_USER), NAME_USER, UcrtVersionNumber.parse(versionDir)));
+                }
+            }
+            return new UcrtFound(getBestUcrt(candidate));
+        } else {
+            return new UcrtNotFound(String.format("The specified installation directory '%s' does not appear to contain a ucrt installation.", candidate));
+        }
+    }
+
+    private SearchResult locateDefaultUcrt() {
+        Ucrt selected = getBestUcrt();
+        return selected == null
+            ? new UcrtNotFound("Could not locate a ucrt installation using the Windows registry.")
+            : new UcrtFound(selected);
+    }
+
+    private static String[] getUcrtVersionDirs(File candidate) {
+        final Pattern ucrtVersionPattern = Pattern.compile("[0-9]+(\\.[0-9]+)*");
+        FileFilter ucrtVersionFilter = new FileFilter() {
+            @Override
+            public boolean accept(File pathname) {
+                Matcher matcher = ucrtVersionPattern.matcher(pathname.getName());
+                if (pathname.isDirectory() && matcher.matches()) {
+                    return true;
+                }
+                return false;
+            }
+        };
+
+        File includeDir = new File(candidate, "Include");
+        File libDir = new File(candidate, "Lib");
+        if (!includeDir.isDirectory() || !libDir.isDirectory()) {
+            return new String[0];
+        }
+
+        Map<String, File> includeDirs = new HashMap<String, File>();
+        for (File dir : includeDir.listFiles(ucrtVersionFilter)) {
+            includeDirs.put(dir.getName(), dir);
+        }
+        Map<String, File> libDirs = new HashMap<String, File>();
+        for (File dir : libDir.listFiles(ucrtVersionFilter)) {
+            libDirs.put(dir.getName(), dir);
+        }
+        Set<String> ucrtVersions = new HashSet<String>();
+        ucrtVersions.addAll(includeDirs.keySet());
+        ucrtVersions.addAll(libDirs.keySet());
+
+        List<String> result = new ArrayList<String>();
+        for (String version : ucrtVersions) {
+            File inc = includeDirs.get(version);
+            File lib = includeDirs.get(version);
```

Comment:
Are both the 'inc' (line no. 158) and 'lib' (line no. 159) variables meant to reference the same Map.Entry<String, File> instance in the 'includeDirs' variable (line no. 144)? I'm guessing that the following was intended:

```java
File inc = includeDirs.get(version);
File lib = libDirs.get(version);
```